<a href="https://colab.research.google.com/github/AnSaradar/World_Cup_History_Analysis/blob/main/World_Cup_History_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Importing Libraries & Loading the Data

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
matches_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/matches.csv')
attendance_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/attendance.csv')
stadiums_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/stadiums.csv')
teams_data = pd.read_csv("/content/drive/MyDrive/Datasets/World_Cup Data/teams.csv")
squads_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/squads.csv')
players_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/players.csv')
tournaments_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/tournaments.csv')
goals_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/goals.csv')
refrees_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/referees_Fiji.csv')
referees_appointments_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/referee_appointments.csv')
referees_appearances_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/referee_appearances.csv')
manageres_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/managers.csv')
manageres_appointments_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/manager_appointments.csv')
managers_appearances_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/manager_appearances.csv')
bookings_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/bookings.csv')
award_winners_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/award_winners.csv')

In [ ]:
!pip install plotly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import plotly.graph_objs as go

#Exploring & Merging the Data

## A- Merging Matches , Teams , Attendance & Stadiums

In [ ]:
attendance_data.shape

(964, 4)

In [ ]:
attendance_data.isnull().sum()

home_team     0
away_team     0
Attendance    0
Date          0
dtype: int64

There are no missing values , and the total number of matches is 964

In [ ]:
teams_data.columns

Index(['key_id', 'team_id', 'team_name', 'team_code', 'federation_name',
       'region_name', 'confederation_id', 'confederation_name',
       'confederation_code', 'team_wikipedia_link',
       'federation_wikipedia_link'],
      dtype='object')

The Key Id Column in the teams dataset is not useful so I will get rid of it

In [ ]:
teams_data.drop(['key_id'],axis=1,inplace=True)

Now  I will Try to merge the attendance with the teams dataset , to get the info of each team like the id and the team code and other info

And Also I will create the match name column from the away and home teams name

In [ ]:
home_teams = teams_data.add_prefix('home_')
away_teams = teams_data.add_prefix('away_')
df= pd.merge(attendance_data, home_teams,how='left' ,left_on='home_team' ,right_on = 'home_team_name')
df= pd.merge(df, away_teams, left_on='away_team' ,right_on = 'away_team_name')
df['match_name'] = df['home_team']+' v '+df['away_team']
df.drop(['home_team_name','away_team_name'],axis=1)
df.shape

(905, 25)

Here after I tried to merge the teams with attendace I notice there are some missings records

In [ ]:
df[df.isnull().any(axis=1)]

home_team              away_team  Attendance        Date  \
71   Korea Republic               Portugal       44097  2022-12-02   
73          IR Iran               Portugal       41685  2018-06-25   
77    Côte d'Ivoire               Portugal       37034  2010-06-15   
144         IR Iran                  Spain       42718  2018-06-20   
157  Korea Republic                  Spain       32733  1990-06-17   
206  Korea Republic                 Poland       48760  2002-06-04   
227         IR Iran          United States       42127  2022-11-29   
238  Korea Republic          United States       60778  2002-06-10   
258      Germany DR              Australia       17000  1974-06-14   
269  Korea Republic                Uruguay       29039  1990-06-21   
290  Korea Republic                Belgium       61397  2014-06-26   
315  Korea Republic                Germany       41835  2018-06-27   
345  Korea Republic                 Mexico       43472  2018-06-23   
359  Korea Republic                 Mexico       39100  1998-06-13   
424  Korea Republic                  Ghana       43983  2022-11-28   
443        China PR             Costa Rica       27217  2002-06-04   
486   Côte d'Ivoire                  Japan       40267  2014-06-14   
521      Germany DR            Netherlands       68348  1974-06-30   
598         IR Iran                Nigeria       39081  2014-06-16   
607  Korea Republic                Algeria       42732  2014-06-22   
616  Korea Republic                 Greece       31513  2010-06-12   
631       Korea DPR                  Chile       13792  1966-07-15   
648  Korea Republic                  Italy       38588  2002-06-18   
653  Korea Republic                  Italy       20000  1986-06-10   
663       Korea DPR                  Italy       17829  1966-07-19   
703         IR Iran                 Angola       38000  2006-06-21   
705   Côte d'Ivoire  Serbia and Montenegro       66000  2006-06-21   
709  Korea Republic                   Togo       48000  2006-06-13   
737  Korea Republic               Bulgaria       45000  1986-06-05   
790  Korea Republic                Bolivia       54453  1994-06-23   
857      Germany DR           West Germany       60200  1974-06-22   

    home_team_id home_team_name home_team_code home_federation_name  \
71           NaN            NaN            NaN                  NaN   
73           NaN            NaN            NaN                  NaN   
77           NaN            NaN            NaN                  NaN   
144          NaN            NaN            NaN                  NaN   
157          NaN            NaN            NaN                  NaN   
206          NaN            NaN            NaN                  NaN   
227          NaN            NaN            NaN                  NaN   
238          NaN            NaN            NaN                  NaN   
258          NaN            NaN            NaN                  NaN   
269          NaN            NaN            NaN                  NaN   
290          NaN            NaN            NaN                  NaN   
315          NaN            NaN            NaN                  NaN   
345          NaN            NaN            NaN                  NaN   
359          NaN            NaN            NaN                  NaN   
424          NaN            NaN            NaN                  NaN   
443          NaN            NaN            NaN                  NaN   
486          NaN            NaN            NaN                  NaN   
521          NaN            NaN            NaN                  NaN   
598          NaN            NaN            NaN                  NaN   
607          NaN            NaN            NaN                  NaN   
616          NaN            NaN            NaN                  NaN   
631          NaN            NaN            NaN                  NaN   
648          NaN            NaN            NaN                  NaN   
653          NaN            NaN            NaN                  N

After Exploring the missing columns , I Notice the problem, there are some team names in the attendace dataset that dont match with the team_name in the teams dataset , and that should be handeled , and the Simple way to handle it is to correct some of teams name

For Example Republic Korea = South Korea , Korea DPR  = North Korea

I will correct the attendace data then I will apply the merge with teams dataframe again |

In [ ]:
attendance_data['home_team'] = attendance_data['home_team'].replace('IR Iran', 'Iran')
attendance_data['away_team'] = attendance_data['away_team'].replace('IR Iran', 'Iran')

attendance_data['home_team'] = attendance_data['home_team'].replace('Korea Republic', 'South Korea')
attendance_data['away_team'] = attendance_data['away_team'].replace('Korea Republic', 'South Korea')

attendance_data['home_team'] = attendance_data['home_team'].replace("Côte d'Ivoire", 'Ivory Coast')
attendance_data['away_team'] = attendance_data['away_team'].replace("Côte d'Ivoire", 'Ivory Coast')

attendance_data['home_team'] = attendance_data['home_team'].replace('Korea DPR', 'North Korea')
attendance_data['away_team'] = attendance_data['away_team'].replace('Korea DPR', 'North Korea')

attendance_data['home_team'] = attendance_data['home_team'].replace('Germany DR', 'East Germany')
attendance_data['away_team'] = attendance_data['away_team'].replace('Germany DR', 'East Germany')

attendance_data['home_team'] = attendance_data['home_team'].replace('China PR', 'China')
attendance_data['away_team'] = attendance_data['away_team'].replace('China PR', 'China')

attendance_data['home_team'] = attendance_data['home_team'].replace('Türkiye', 'Turkey')
attendance_data['away_team'] = attendance_data['away_team'].replace('Türkiye', 'Turkey')

attendance_data['home_team'] = attendance_data['home_team'].replace('FR Yugoslavia', 'Yugoslavia')
attendance_data['away_team'] = attendance_data['away_team'].replace('FR Yugoslavia', 'Yugoslavia')


In [ ]:
home_teams = teams_data.add_prefix('home_')
away_teams = teams_data.add_prefix('away_')
df= pd.merge(attendance_data, home_teams,how='left' ,left_on='home_team' ,right_on = 'home_team_name')
df= pd.merge(df, away_teams, left_on='away_team' ,right_on = 'away_team_name')
df['match_name'] = df['home_team']+' v '+df['away_team']
df.drop(['home_team_name','away_team_name'],axis=1,inplace=True)
df.shape

(964, 23)

In [ ]:
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [home_team, away_team, Attendance, Date, home_team_id, home_team_code, home_federation_name, home_region_name, home_confederation_id, home_confederation_name, home_confederation_code, home_team_wikipedia_link, home_federation_wikipedia_link, away_team_id, away_team_code, away_federation_name, away_region_name, away_confederation_id, away_confederation_name, away_confederation_code, away_team_wikipedia_link, away_federation_wikipedia_link, match_name]
Index: []

[0 rows x 23 columns]

Now the Problem is solved

In [ ]:
df.columns

Index(['home_team', 'away_team', 'Attendance', 'Date', 'home_team_id',
       'home_team_code', 'home_federation_name', 'home_region_name',
       'home_confederation_id', 'home_confederation_name',
       'home_confederation_code', 'home_team_wikipedia_link',
       'home_federation_wikipedia_link', 'away_team_id', 'away_team_code',
       'away_federation_name', 'away_region_name', 'away_confederation_id',
       'away_confederation_name', 'away_confederation_code',
       'away_team_wikipedia_link', 'away_federation_wikipedia_link',
       'match_name'],
      dtype='object')

In [ ]:
df.rename(columns={'Date':'match_date'},inplace=True)

In [ ]:
df.columns

Index(['home_team', 'away_team', 'Attendance', 'match_date', 'home_team_id',
       'home_team_code', 'home_federation_name', 'home_region_name',
       'home_confederation_id', 'home_confederation_name',
       'home_confederation_code', 'home_team_wikipedia_link',
       'home_federation_wikipedia_link', 'away_team_id', 'away_team_code',
       'away_federation_name', 'away_region_name', 'away_confederation_id',
       'away_confederation_name', 'away_confederation_code',
       'away_team_wikipedia_link', 'away_federation_wikipedia_link',
       'match_name'],
      dtype='object')

In [ ]:
df.to_csv('df.csv')

So I Solved the problem and Now there are no missing values

Now I will Merge the matches dataframe with the stadiums dataframe to get all the info of the stadiums

In [ ]:
matches_data.shape

(964, 37)

Notice that the number of records is the same in each teams and matches

In [ ]:
matches_data.drop(['key_id'],axis=1,inplace=True)
matches_data.columns

Index(['tournament_id', 'tournament_name', 'match_id', 'match_name',
       'stage_name', 'group_name', 'group_stage', 'knockout_stage', 'replayed',
       'replay', 'match_date', 'match_time', 'stadium_id', 'stadium_name',
       'city_name', 'country_name', 'home_team_id', 'home_team_name',
       'home_team_code', 'away_team_id', 'away_team_name', 'away_team_code',
       'score', 'home_team_score', 'away_team_score', 'home_team_score_margin',
       'away_team_score_margin', 'extra_time', 'penalty_shootout',
       'score_penalties', 'home_team_score_penalties',
       'away_team_score_penalties', 'result', 'home_team_win', 'away_team_win',
       'draw'],
      dtype='object')

In [ ]:
stadiums_data.drop(['key_id'],axis=1,inplace=True)


I will merge these two dataframes , But Notice there are some duplicates features , So we have to drop them before the merge

In [ ]:
matches_data = pd.merge(matches_data,stadiums_data.drop(['stadium_name','city_name','country_name'],axis=1),on = 'stadium_id',how='left')
matches_data.shape

(964, 39)

In [ ]:
matches_data.columns

Index(['tournament_id', 'tournament_name', 'match_id', 'match_name',
       'stage_name', 'group_name', 'group_stage', 'knockout_stage', 'replayed',
       'replay', 'match_date', 'match_time', 'stadium_id', 'stadium_name',
       'city_name', 'country_name', 'home_team_id', 'home_team_name',
       'home_team_code', 'away_team_id', 'away_team_name', 'away_team_code',
       'score', 'home_team_score', 'away_team_score', 'home_team_score_margin',
       'away_team_score_margin', 'extra_time', 'penalty_shootout',
       'score_penalties', 'home_team_score_penalties',
       'away_team_score_penalties', 'result', 'home_team_win', 'away_team_win',
       'draw', 'stadium_capacity', 'stadium_wikipedia_link',
       'city_wikipedia_link'],
      dtype='object')

Now it last step, which it is merging the matches with the attendance to get the attendance for each match

In [ ]:
df.columns

Index(['home_team', 'away_team', 'Attendance', 'match_date', 'home_team_id',
       'home_team_code', 'home_federation_name', 'home_region_name',
       'home_confederation_id', 'home_confederation_name',
       'home_confederation_code', 'home_team_wikipedia_link',
       'home_federation_wikipedia_link', 'away_team_id', 'away_team_code',
       'away_federation_name', 'away_region_name', 'away_confederation_id',
       'away_confederation_name', 'away_confederation_code',
       'away_team_wikipedia_link', 'away_federation_wikipedia_link',
       'match_name'],
      dtype='object')

In [ ]:
def sort_string(x):
  l = x.split()
  l.sort()
  st = ' '.join(l)
  return st

I will merge on home/away team code after I sort them to avoid swapping the data, and also on match_date

In [ ]:
matches_data['code_pair'] = matches_data['home_team_code'] +' '+ matches_data['away_team_code']
matches_data['sorted_code_pair'] = matches_data['code_pair'].apply(lambda x : sort_string(x))

df['code_pair'] = df['home_team_code'] +' '+ df['away_team_code']
df['sorted_code_pair'] = df['code_pair'].apply(lambda x : sort_string(x))

In [ ]:
matches_data = pd.merge(matches_data,df.drop(['home_team','away_team','home_team_id','away_team_id','match_name'],axis=1),on=['sorted_code_pair','match_date'],how='left')

In [ ]:
matches_data[matches_data.isnull().any(axis=1)]

Empty DataFrame
Columns: [tournament_id, tournament_name, match_id, match_name, stage_name, group_name, group_stage, knockout_stage, replayed, replay, match_date, match_time, stadium_id, stadium_name, city_name, country_name, home_team_id, home_team_name, home_team_code_x, away_team_id, away_team_name, away_team_code_x, score, home_team_score, away_team_score, home_team_score_margin, away_team_score_margin, extra_time, penalty_shootout, score_penalties, home_team_score_penalties, away_team_score_penalties, result, home_team_win, away_team_win, draw, stadium_capacity, stadium_wikipedia_link, city_wikipedia_link, code_pair_x, sorted_code_pair, Attendance, home_team_code_y, home_federation_name, home_region_name, home_confederation_id, home_confederation_name, home_confederation_code, home_team_wikipedia_link, home_federation_wikipedia_link, away_team_code_y, away_federation_name, away_region_name, away_confederation_id, away_confederation_name, away_confederation_code, away_team_wikipedia_link, away_federation_wikipedia_link, code_pair_y]
Index: []

[0 rows x 59 columns]

So I Solved the problem from the previous task by creating the sorted_pair_code

##B- Players Names & Squads

In [ ]:
def get_teams_names_played_for(player_id):
    player_squads = squads_data[squads_data['player_id'] == player_id]
    teams_played_for = set(player_squads['team_name'].unique())

    return teams_played_for

In [ ]:
def get_teams_codes_played_for(player_id):
    player_squads = squads_data[squads_data['player_id'] == player_id]
    teams_played_for = set(player_squads['team_code'].unique())

    return teams_played_for

In [ ]:
def get_num_of_teams__played_for(player_id):
    player_squads = squads_data[squads_data['player_id'] == player_id]
    teams_played_for = len(set(player_squads['team_name'].unique()))

    return teams_played_for

In [ ]:
players_data['names_of_teams_played_for'] = players_data['player_id'].apply(get_teams_names_played_for)
players_data['codes_of_teams_played_for'] = players_data['player_id'].apply(get_teams_codes_played_for)
players_data['num_of_teams_played_for'] = players_data['player_id'].apply(get_num_of_teams__played_for)


In [ ]:
players_teams = players_data
players_teams.head()

key_id player_id family_name given_name  birth_date  goal_keeper  defender  \
0       1   P-08891     A'Court       Alan  1934-09-30            0         0   
1       2   P-08589    Aaronson    Brenden  2000-10-22            0         0   
2       3   P-04897   Abadzhiev     Stefan  1934-07-03            0         0   
3       4   P-05556       Abalo  Jean-Paul  1975-06-26            0         1   
4       5   P-08163      Abanda    Patrice  1978-08-03            0         1   

   midfielder  forward  count_tournaments list_tournaments  \
0           0        1                  1             1958   
1           0        1                  1             2022   
2           0        1                  1             1966   
3           0        0                  1             2006   
4           0        0                  1             1998   

                            player_wikipedia_link names_of_teams_played_for  \
0    https://en.wikipedia.org/wiki/Alan_A%27Court                 {England}   
1  https://en.wikipedia.org/wiki/Brenden_Aaronson           {United States}   
2  https://en.wikipedia.org/wiki/Stefan_Abadzhiev                {Bulgaria}   
3   https://en.wikipedia.org/wiki/Jean-Paul_Abalo                    {Togo}   
4    https://en.wikipedia.org/wiki/Patrice_Abanda                {Cameroon}   

  codes_of_teams_played_for  num_of_teams_played_for  
0                     {ENG}                        1  
1                     {USA}                        1  
2                     {BGR}                        1  
3                     {TGO}                        1  
4                     {CMR}                        1

In [ ]:
players_teams[players_teams.isnull().any(axis=1)]

key_id player_id family_name given_name birth_date  goal_keeper  \
7061    7062   P-08148   Sepúlveda  Guillermo        NaN            0   

      defender  midfielder  forward  count_tournaments list_tournaments  \
7061         1           1        0                  2       1958, 1962   

                                  player_wikipedia_link  \
7061  https://en.wikipedia.org/wiki/Guillermo_Sep%C3...   

     names_of_teams_played_for codes_of_teams_played_for  \
7061                  {Mexico}                     {MEX}   

      num_of_teams_played_for  
7061                        1

In [ ]:
players_teams.at[7061,'birth_date']='1934-2-28'

Getting his birthdate from wiki

In [ ]:
players_teams[players_teams.isnull().any(axis=1)]

Empty DataFrame
Columns: [key_id, player_id, family_name, given_name, birth_date, goal_keeper, defender, midfielder, forward, count_tournaments, list_tournaments, player_wikipedia_link, names_of_teams_played_for, codes_of_teams_played_for, num_of_teams_played_for]
Index: []

#Feature Engineering

This stage will be Focuse on creating a new features

- Total Goals in match

In [ ]:
matches_data['total_goals_in_match'] = matches_data['home_team_score'] + matches_data['away_team_score']

- Match For Host (if the Host Country is playing in this match)

In [ ]:
matches_data['match_for_host'] = ((matches_data['home_team_name'] == matches_data['country_name']) | (matches_data['away_team_name'] == matches_data['country_name'])).astype(int)


- Used of Stadium Capacity

In [ ]:
matches_data['ratio_capacity_used'] = matches_data['Attendance'] / matches_data['stadium_capacity']

In [ ]:
matches_data['ratio_capacity_used'].describe()

count    964.000000
mean       0.831134
std        0.229490
min        0.022222
25%        0.722222
50%        0.924674
75%        0.990742
max        2.112903
Name: ratio_capacity_used, dtype: float64

- Attendance_Category & Relative Attendance Category

For the Attendance Category

In [ ]:
matches_data['Attendance'].describe()

count       964.000000
mean      45693.374481
std       22704.131689
min        2000.000000
25%       31800.000000
50%       42725.000000
75%       60984.500000
max      173850.000000
Name: Attendance, dtype: float64

So I will make 6 Categories {Low , Normal , High , Very high ,Crowded , Over Crowded}

In [ ]:
def categorize_attendance(attendance):
    if attendance < 5000:
        return 'Low'
    elif attendance < 20000:
        return 'Normal'
    elif attendance < 50000:
        return 'High'
    elif attendance < 100000:
        return 'Very high'
    elif attendance < 150000:
        return 'Crowded'
    else:
        return 'Over Crowded'

In [ ]:
matches_data['attendance_category'] = matches_data['Attendance'].apply(categorize_attendance)

For Relative Attendance Category

In [ ]:
def get_percentage_category(percentage):
    percentage = percentage * 100
    if percentage < 20:
        return 'Low'
    elif 20 <= percentage < 40:
        return 'Normal'
    elif 40 <= percentage < 60:
        return 'High'
    elif 60 <= percentage < 80:
        return 'Very High'
    elif 80 <= percentage < 95:
        return 'Crowded'
    else:
        return 'Over Crowded'

In [ ]:
matches_data['relative_attendance_category'] = matches_data['Attendance'].apply(categorize_attendance)
matches_data['relative_attendance_category'].head()

0       Low
1    Normal
2      High
3       Low
4      High
Name: relative_attendance_category, dtype: object

- Host Country Code

In [ ]:
def get_team_code(team_name):
    team_code = teams_data.loc[teams_data['team_name'] == team_name, 'team_code'].iloc[0]
    return team_code

In [ ]:
matches_data['host_country_code'] = matches_data['country_name'].apply(get_team_code)

- Tournament Year

In [ ]:
matches_data['match_date'] = pd.to_datetime(matches_data['match_date'])
matches_data['tournament_year'] = matches_data['match_date'].dt.year

- Player Full Name

In [ ]:
players_data['given_name'].replace('not applicable', '', inplace=True)
players_data['family_name'].replace('not applicable', '', inplace=True)

In [ ]:
players_data['full_name'] = players_data['given_name'] +' '+ players_data['family_name']

- Short Stage Name

In [ ]:
matches_data['short_stage_name'] = matches_data['stage_name'].apply(lambda x: 'group_stage' if x in ['group stage','second group stage'] else 'knockout_stage')

- Winner Code

In [ ]:
def get_team_code(team_name):

    matching_team = teams_data[teams_data['team_name'] == team_name]

    return matching_team.iloc[0]['team_code']

In [ ]:
tournaments_data['winner_code'] = tournaments_data['winner'].apply(get_team_code)

- Late Goal

For me the late goals are the Goals that have been scored in the last 5 mins of the Second_half OR Extra_time Second_half

In [ ]:
def is_late_goal(goal):
    if goal['match_period'] == 'second half' and goal['minute_regulation'] >= 85:
        return 1
    elif goal['match_period'] == 'extra time, second half' and goal['minute_regulation'] >= 115:
        return 1
    else:
        return 0

In [ ]:
goals_data['late_goal'] = goals_data.apply(is_late_goal, axis=1)

- Number of Goals for each player

In [ ]:
def get_player_goals(player_id):
    player_goals_df = goals_data[(goals_data['player_id'] == player_id) & (goals_data['own_goal'] == 0)]
    player_goals = player_goals_df['goal_id'].count()

    return player_goals

In [ ]:
players_teams['goals_scored'] = players_teams['player_id'].apply(get_player_goals)

In [ ]:
sorted = players_teams.sort_values('goals_scored', ascending=False)
sorted.head()

key_id player_id family_name given_name  birth_date  goal_keeper  \
4135    4136   P-05224       Klose   Miroslav  1978-06-09            0   
6691    6692   P-08490     Ronaldo             1976-09-18            0   
5411    5412   P-02173      Müller       Gerd  1945-11-03            0   
5133    5134   P-03429       Messi     Lionel  1987-06-24            0   
2674    2675   P-02537    Fontaine       Just  1933-08-18            0   

      defender  midfielder  forward  count_tournaments  \
4135         0           0        1                  4   
6691         0           0        1                  4   
5411         0           0        1                  2   
5133         0           0        1                  5   
2674         0           0        1                  1   

                  list_tournaments  \
4135        2002, 2006, 2010, 2014   
6691        1994, 1998, 2002, 2006   
5411                    1970, 1974   
5133  2006, 2010, 2014, 2018, 2022   
2674                          1958   

                                  player_wikipedia_link  \
4135       https://en.wikipedia.org/wiki/Miroslav_Klose   
6691  https://en.wikipedia.org/wiki/Ronaldo_(Brazili...   
5411     https://en.wikipedia.org/wiki/Gerd_M%C3%BCller   
5133         https://en.wikipedia.org/wiki/Lionel_Messi   
2674        https://en.wikipedia.org/wiki/Just_Fontaine   

     names_of_teams_played_for codes_of_teams_played_for  \
4135                 {Germany}                     {DEU}   
6691                  {Brazil}                     {BRA}   
5411            {West Germany}                     {DEU}   
5133               {Argentina}                     {ARG}   
2674                  {France}                     {FRA}   

      num_of_teams_played_for       full_name  goals_scored  
4135                        1  Miroslav Klose            16  
6691                        1         Ronaldo            15  
5411                        1     Gerd Müller            14  
5133                        1    Lionel Messi            13  
2674                        1   Just Fontaine            13

#Mining & Analysis

In [ ]:
import plotly.express as px

## A- Attendance

###MEAN , Median  Attendance of World Cup Matches

In [ ]:
mean_attendance = matches_data.groupby("tournament_name")["Attendance"].mean().reset_index()
median_attendance = matches_data.groupby("tournament_name")["Attendance"].median().reset_index()

In [ ]:
fig = px.line(title="Mean & Median of the World Cup matches Attendance Over the Years")
fig.add_scatter(x=mean_attendance["tournament_name"], y=mean_attendance["Attendance"], mode='lines', name='Mean Attendance')
fig.add_scatter(x=median_attendance["tournament_name"], y=median_attendance["Attendance"], mode='lines', name='Median Attendance')
fig.update_xaxes(tickangle=45)
fig.show()

We Notice that the USA World Cup in 1994 is the best world cup for number of Audiance

###Attendance Histogram

In [ ]:
fig = px.histogram(matches_data, x="Attendance", nbins=100, title="Attendance Histogram")
fig.show()

We Notice that the median value of attendance is about 42k , and there are very few matches where the number of attendace exceeded 100K.
And the most Crowded game was about 170K.

###Attendance Boxplot for each Tournaments

In [ ]:
fig = px.box(matches_data, x="tournament_name", y="Attendance", title="Attendance Box Plot by Year")
fig.update_xaxes(tickangle=45)
fig.show()

The max number of attendance was at The 1950 world cup .
the min number of attendacne was at the first world cup in 1930.
The largest median of attendace was at the 1994 world cup.


##B- Goals

###Mean min goal

In [ ]:
mean_match_period_by_tournament = goals_data.groupby('tournament_name')['minute_regulation'].mean()

fig = px.bar(mean_match_period_by_tournament, x=mean_match_period_by_tournament.index, y=mean_match_period_by_tournament.values,
             labels={'x': 'Tournament', 'y': 'Mean Match Period'})
fig.show()

###Histogram for Total number of goals per Match

In [ ]:
total_goals_per_match = goals_data.groupby('match_id')['goal_id'].count()

fig = go.Figure(data=[go.Histogram(x=total_goals_per_match, nbinsx=20)])

fig.update_layout(title='Distribution of Total Goals per Match',
                  xaxis_title='Total Goals',
                  yaxis_title='Count')
fig.show()

###Most Frequant min for each tournaments

In [ ]:
most_frequent_minute = goals_data.groupby('tournament_name')['minute_regulation'].apply(lambda x: x.mode().values[0]).reset_index()

fig = px.bar(most_frequent_minute, x='tournament_name', y='minute_regulation',
             title='Most Frequent Minute Regulation for Each Tournament')
fig.update_layout(xaxis_title='Tournament Name', yaxis_title='Most Frequent Minute Regulation')


fig.show()

###Histogram for late goals for all tournamnets

In [ ]:
late_goals = goals_data[goals_data['late_goal'] == True]
late_goals_by_tournament = late_goals.groupby('tournament_name')['goal_id'].count()

fig = px.histogram(late_goals_by_tournament, x=late_goals_by_tournament.values, nbins=20, title='Distribution of Late Goals by Tournament')

fig.update_layout(xaxis_title='Number of Late Goals',
                  yaxis_title='Count')

fig.show()

###Top Scorers of all Time

In [ ]:
goals_data['given_name'].replace('not applicable', '', inplace=True)
goals_data['family_name'].replace('not applicable', '', inplace=True)

goals_data['full_name'] = goals_data['given_name']+' '+goals_data['family_name']

In [ ]:
non_own_goals = goals_data[goals_data['own_goal'] == False]

player_goals = non_own_goals.groupby('full_name')['goal_id'].count().reset_index()
player_goals = player_goals.sort_values(by='goal_id', ascending=False)

top_scorers = player_goals[:12]

fig = px.bar(top_scorers, x='full_name', y='goal_id', title='Top Scorers in The World Cup History ')


fig.update_layout(xaxis_title='Player',
                  yaxis_title='Number of Goals')

###Top Scorers for each Tournaments

In [ ]:
no_own_goals = goals_data[goals_data['own_goal']==0]

scorers = no_own_goals.groupby(['tournament_id', 'player_id', 'full_name'])['goal_id'].count().reset_index()
scorers = scorers.rename(columns={'goal_id': 'total_goals'})

max_goals = scorers.groupby('tournament_id')['total_goals'].max().reset_index()

top_scorers = scorers.merge(max_goals, on=['tournament_id', 'total_goals'])

top_scorers = top_scorers.sort_values(['tournament_id', 'total_goals'], ascending=[True, False])

fig = px.bar(top_scorers, x='tournament_id', y='total_goals', color='full_name',
             labels={'tournament_id': 'Tournament ID', 'total_goals': 'Total Goals', 'full_name': 'Top Scorer'},
             title='Top Scorers for Each Tournament')
fig.layout.update(showlegend=False)
fig.show()

###Total Goals Scored in each Tournament

In [ ]:
goals_count = goals_data.groupby('tournament_name')['goal_id'].count().reset_index()
goals_count = goals_count.rename(columns={'goal_id': 'goals_count'})

#goals_count = goals_count.sort_values('goals_count', ascending=False)

fig = px.bar(goals_count, x='goals_count', y='tournament_name',
             labels={'goals_count': 'Goals Count', 'tournament_name': 'Tournament Name'},
             title='Number of Goals Scored in Each Tournament')
fig.show()

Most of the top tournaments by goals are recent tournaments , and it makes sense because in the past the number of the participated teams was less so the number of matches for a tournament was less

###Strip plot for goal_min & stage_name

In [ ]:
goals_data['short_stage_name'] = goals_data['stage_name'].apply(lambda x: 'group_stage' if x.lower() in ['group stage', 'second group stage'] else 'knockout_stage')

In [ ]:
filtered_goals = goals_data[(goals_data['own_goal'] == False) & (goals_data['team_name'].isin(['Brazil', 'Germany', 'Italy']))]

# Create a strip plot
fig = px.strip(filtered_goals, x='short_stage_name', y='minute_label', color='team_name')

# Customize the plot
fig.update_layout(
    title='Distribution of Goal Minutes by Stage Name',
    xaxis_title='Stage Name',
    yaxis_title='Goal Minute'
)

# Display the plot
fig.show()

##C-Matches

In [ ]:
matches_data['home_team_name_no_spaces'] = matches_data['home_team_name'].str.replace(' ', '')
matches_data['away_team_name_no_spaces'] = matches_data['away_team_name'].str.replace(' ', '')
matches_data['team_pair'] = matches_data['home_team_name_no_spaces']+' '+matches_data['away_team_name_no_spaces']

In [ ]:
def sort_string(x):
  l = x.split()
  l.sort()
  st = ' '.join(l)
  return st

In [ ]:
matches_data['sorted_team_pair'] = matches_data['team_pair'].apply(lambda x: sort_string(x))


In [ ]:
value_counts = matches_data['sorted_team_pair'].value_counts()

N = 10
fig = px.bar(value_counts[:N], x=value_counts[:N].index, y=value_counts[:N], labels={'x': 'Values', 'y': 'Frequency'}, title=f'Top {N} Most Frequent Matches in The World Cup History')
fig.update_xaxes(tickangle=45)
fig.show()

##D-Touraments

In [ ]:
tournamnets_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/tournaments.csv')


###Players Played for multiple teams

In [ ]:
players_played_for_different_teams = players_data[players_data['num_of_teams_played_for']>1]
players_played_for_different_teams

key_id player_id  family_name  given_name  birth_date  goal_keeper  \
331      332   P-01512     Altafini        José  1938-07-24            0   
937      938   P-01675     Berthold      Thomas  1964-11-12            0   
1039    1040   P-03207       Bokšić        Alen  1970-01-21            0   
1087    1088   P-05308     Borodyuk   Aleksandr  1962-11-30            0   
1166    1167   P-00830       Brehme     Andreas  1960-11-09            0   
1227    1228   P-05379     Buchwald       Guido  1961-01-24            0   
2037    2038   P-02845      Demaría     Attilio  1909-03-19            0   
3036    3037   P-06666  Gorlukovich      Sergei  1961-11-18            0   
3294    3295   P-02040       Häßler      Thomas  1966-05-30            0   
3572    3573   P-04641      Illgner        Bodo  1967-04-07            1   
3719    3720   P-04605        Jarni      Robert  1968-10-26            0   
4131    4132   P-06604    Klinsmann      Jürgen  1964-07-30            0   
4155    4156   P-08277       Kohler      Jürgen  1965-10-06            0   
4200    4201   P-03556        Köpke     Andreas  1962-03-12            1   
4968    4969   P-09502     Matthäus      Lothar  1961-03-21            0   
5200    5201   P-09293    Milošević        Savo  1973-09-02            0   
5280    5281   P-01103       Möller     Andreas  1967-09-02            0   
5304    5305   P-02543        Monti        Luis  1901-05-15            0   
6319    6320   P-02576   Prosinečki      Robert  1969-01-12            0   
6340    6341   P-01259       Puskás      Ferenc  1927-04-02            0   
6396    6397   P-01757        Raftl      Rudolf  1911-02-07            1   
6506    6507   P-06121       Reuter      Stefan  1966-10-16            0   
6545    6546   P-07779       Riedle  Karl-Heinz  1965-09-16            0   
6903    6904   P-05394   Santamaría        José  1929-07-31            0   
6974    6975   P-09633      Schmaus   Willibald  1911-06-16            0   
7353    7354   P-02369    Stanković       Dejan  1978-09-11            0   
7403    7404   P-08793    Stojković    Vladimir  1983-07-28            1   
7424    7425   P-08975        Stroh       Josef  1913-03-05            0   
7451    7452   P-01739        Šuker       Davor  1968-01-01            0   
7620    7621   P-07346         Thon        Olaf  1966-05-01            0   
8030    8031   P-06299        Vidić     Nemanja  1981-10-21            0   
8093    8094   P-02452       Völler        Rudi  1960-04-13            0   
8131    8132   P-00537       Wagner       Franz  1911-09-23            0   
8443    8444   P-02502        Žigić      Nikola  1980-09-25            0   

      defender  midfielder  forward  count_tournaments  \
331          0           0        1                  2   
937          1           0        0                  3   
1039         0           0        1                  2   
1087         0           1        0                  2   
1166         1           0        0                  3   
1227         1           0        0                  2   
2037         0           0        1                  2   
3036         1           0        0                  2   
3294         0           1        0                  3   
3572         0           0        0                  2   
3719         1           1        0                  3   
4131         0           0        1                  3   
4155         1           0        0                  3   
4200         0           0        0                  3   
4968         1           1        0                  5   
5200         0           0        1                  2   
5280         0           1        0                  3   
5304         0           1        0                  2   
6319         0           1        0                  3   
6340         0           0        1                  2   
6396         0           0        0                  2   
6506         1           0        0                  2   
6545       

Most of the players played for multiple teams for politics reasons , and Wars

###Corrleation/Association Between Host & Winner

In [ ]:
from scipy.stats import chi2_contingency

In [ ]:
contingency_table = pd.crosstab(tournamnets_data['host_country'], tournamnets_data['winner'])

_, p_value, _, _ = chi2_contingency(contingency_table)

chi2_stat = chi2_contingency(contingency_table)[0]
n = contingency_table.sum().sum()
min_dim = min(contingency_table.shape)

cramers_v = np.sqrt(chi2_stat / (n * (min_dim - 1)))


print("Cramer's V:", cramers_v)
print("p-value:", p_value)

Cramer's V: 0.8885850175044215
p-value: 0.4133433398768274


Their is a String Association between the host and the winner , and it makes sense because the host was the winner 5 times out of 23 World cup so it is about 20% of times

###Association between Host Match & Relative Attendance Category

In [ ]:
contingency_table = pd.crosstab(matches_data['match_for_host'],matches_data['relative_attendance_category'])

chi2_stat = chi2_contingency(contingency_table)[0]

n = contingency_table.sum().sum()
min_dim = min(contingency_table.shape)

cramers_v = np.sqrt(chi2_stat / (n * (min_dim - 1)))

print("Cramer's V:", cramers_v)

Cramer's V: 0.31216107362445555


###Association between Host Match & Attendance Category

In [ ]:
contingency_table = pd.crosstab(matches_data['match_for_host'],matches_data['attendance_category'])

chi2_stat = chi2_contingency(contingency_table)[0]

n = contingency_table.sum().sum()
min_dim = min(contingency_table.shape)

cramers_v = np.sqrt(chi2_stat / (n * (min_dim - 1)))

print("Cramer's V:", cramers_v)

Cramer's V: 0.31216107362445555


We Notcie that their is a moderate association between host Match and both{Attendance Category,Relative Attendance Category}

#Secret Mission

##Refrees

In [ ]:

refrees_data.head()

key_id referee_id       family_name given_name      country_name  \
0       1      R-001       Abdel-Fatah      Essam             Egypt   
1       2      R-002             Adair       John  Northern Ireland   
2       3      R-003           Agnolin      Luigi             Italy   
3       4      R-004           Aguilar       Joel       El Salvador   
4       5      R-005  Aguilar Elizalde       Abel            Mexico   

  confederation_id                                 confederation_name  \
0             CF-2                  Confederation of African Football   
1             CF-6            Union of European Football Associations   
2             CF-6            Union of European Football Associations   
3             CF-3  Confederation of North, Central American and C...   
4             CF-3  Confederation of North, Central American and C...   

  confederation_code                            referee_wikipedia_link  
0                CAF  https://en.wikipedia.org/wiki/Essam_Abd_El_Fatah  
1               UEFA                                     not available  
2               UEFA       https://en.wikipedia.org/wiki/Luigi_Agnolin  
3           CONCACAF        https://en.wikipedia.org/wiki/Joel_Aguilar  
4           CONCACAF                                     not available

###Referees Nationality

In [ ]:
referee_counts = refrees_data['country_name'].value_counts().reset_index()
referee_counts.columns = ['country_name', 'Count']

fig = px.bar(referee_counts, x='country_name', y='Count', labels={'country_name': 'Country', 'Count': 'Number of Referees'}, title='Number of Referees from Each Country')

fig.show()

###Most Apperances Referees (By the number of matches)

In [ ]:
referees_appearances_data['full_name'] = referees_appearances_data['given_name']+' '+referees_appearances_data['family_name']


In [ ]:
referee_counts = referees_appearances_data['full_name'].value_counts().reset_index()
referee_counts.columns = ['Referee', 'Count']
referee_counts = referee_counts[0:10]
# Create a bar plot
fig = px.bar(referee_counts, x='Referee', y='Count', labels={'Referee': 'Referee', 'Count': 'Apperance'}, title='Top 10 Referees Appearnace in the World Cup History')

# Show the plot
fig.show()

###Most Apperances Referees (By the number of Tournaments)

In [ ]:
referees_appointments_data['full_name'] = referees_appointments_data['given_name']+' '+referees_appointments_data['family_name']


In [ ]:
referee_unique_tournaments = referees_appointments_data.groupby('full_name')['tournament_name'].nunique().reset_index()
referee_unique_tournaments.columns = ['Referee', 'Unique Tournaments']
referee_unique_tournaments = referee_unique_tournaments.sort_values('Unique Tournaments', ascending=False)
top_20_referees = referee_unique_tournaments.head(20)
fig = px.bar(top_20_referees, x='Referee', y='Unique Tournaments', labels={'Referee': 'Referee', 'Unique Tournaments': 'Number of Unique Tournaments'}, title='Number of Unique Tournaments for Each Referee')


fig.show()

##Managers

In [ ]:
manageres_data.head()

key_id manager_id family_name given_name country_name  \
0       1      M-001      Acosta     Nelson      Uruguay   
1       2      M-002        Addo       Otto        Ghana   
2       3      M-003     Adshead       John      England   
3       4      M-004    Advocaat       Dick  Netherlands   
4       5      M-005     Aguirre     Javier       Mexico   

                         manager_wikipedia_link  
0   https://en.wikipedia.org/wiki/Nelson_Acosta  
1       https://en.wikipedia.org/wiki/Otto_Addo  
2    https://en.wikipedia.org/wiki/John_Adshead  
3   https://en.wikipedia.org/wiki/Dick_Advocaat  
4  https://en.wikipedia.org/wiki/Javier_Aguirre

###Managers Nationality

In [ ]:
manageres_data['full_name'] = manageres_data['given_name'] +' '+manageres_data['family_name']

In [ ]:
manager_counts = manageres_data['country_name'].value_counts().reset_index()
manager_counts.columns = ['country_name', 'Count']
manager_counts = manager_counts[:10]
fig = px.bar(manager_counts, x='country_name', y='Count', labels={'country_name': 'Country', 'Count': 'Number of Referees'}, title='Number of Referees from Each Country')

fig.show()

###Is their any asscosiation between the manager managing his home team and winning the trophy

In [ ]:
winner_manager_nationality = pd.merge(tournamnets_data, manageres_appointments_data, left_on=['tournament_id', 'tournament_name','winner'],right_on=['tournament_id', 'tournament_name','team_name'], how='left')
winner_manager_nationality['national_manager'] = winner_manager_nationality.apply(lambda row: 1 if row['country_name'] == row['winner'] else 0, axis=1)

winner_manager_nationality.head()

key_id_x tournament_id      tournament_name  year  start_date    end_date  \
0         1       WC-1930  1930 FIFA World Cup  1930  1930-07-13  1930-07-30   
1         2       WC-1934  1934 FIFA World Cup  1934  1934-05-27  1934-06-10   
2         3       WC-1938  1938 FIFA World Cup  1938  1938-06-04  1938-06-19   
3         4       WC-1950  1950 FIFA World Cup  1950  1950-06-24  1950-07-16   
4         5       WC-1954  1954 FIFA World Cup  1954  1954-06-16  1954-07-04   

  host_country        winner  host_won  count_teams  ...  final  key_id_y  \
0      Uruguay       Uruguay         1           13  ...      1        13   
1        Italy         Italy         1           16  ...      1        28   
2       France         Italy         0           15  ...      1        43   
3       Brazil       Uruguay         0           13  ...      0        56   
4  Switzerland  West Germany         0           16  ...      1        65   

   team_id     team_name  team_code  manager_id  family_name  given_name  \
0     T-81       Uruguay        URY       M-331      Suppici     Alberto   
1     T-39         Italy        ITA       M-266        Pozzo    Vittorio   
2     T-39         Italy        ITA       M-266        Pozzo    Vittorio   
3     T-81       Uruguay        URY       M-180        López        Juan   
4     T-83  West Germany        DEU       M-121    Herberger        Sepp   

   country_name national_manager  
0       Uruguay                1  
1         Italy                1  
2         Italy                1  
3       Uruguay                1  
4       Germany                0  

[5 rows x 27 columns]

In [ ]:
contingency_table = pd.crosstab(winner_manager_nationality['winner'],winner_manager_nationality['national_manager'])

chi2_stat = chi2_contingency(contingency_table)[0]

n = contingency_table.sum().sum()
min_dim = min(contingency_table.shape)

cramers_v = np.sqrt(chi2_stat / (n * (min_dim - 1)))

print("Cramer's V:", cramers_v)

Cramer's V: 0.9999999999999999


And it makes sense because most of the winner teams have a local manager

##Top Teams Evaluation

###Rank by Title

In [ ]:
tournaments_data['winner'] = tournaments_data['winner'].replace('West Germany', 'Germany')
team_winner_count = tournaments_data['winner'].value_counts()
fig = px.bar(x=team_winner_count.index, y=team_winner_count.values, title='Count of Teams that Won Tournaments')

fig.update_layout(xaxis_title='Team', yaxis_title='Count')

fig.show()

###Rank by the number of goals

In [ ]:
goals_data['team_name'] = goals_data['team_name'].replace('West Germany', 'Germany')

non_own_goals = goals_data[goals_data['own_goal'] == False]

team_goal_count = non_own_goals.groupby('team_name')['goal_id'].count().reset_index()
team_goal_count.loc[team_goal_count['team_name'].isin(['Germany', 'Germany/West Germany']), 'team_name'] = 'Germany/West Germany'
team_goal_count = team_goal_count.sort_values('goal_id', ascending=False)

fig = px.bar(team_goal_count, x='team_name', y='goal_id', title='Number of Goals (Excluding Own Goals) by Team')
fig.update_layout(xaxis_title='Team', yaxis_title='Goal Count')

fig.show()

###Rank by the number of Matches

In [ ]:
qualifications_data = pd.read_csv('/content/drive/MyDrive/Datasets/World_Cup Data/qualified_teams.csv')
qualifications_data.columns

Index(['key_id', 'tournament_id', 'tournament_name', 'team_id', 'team_name',
       'team_code', 'count_matches', 'performance'],
      dtype='object')

In [ ]:
team_appearances = qualifications_data.groupby('team_name')['count_matches'].sum().reset_index()

team_appearances = team_appearances.sort_values('count_matches', ascending=False)

fig = px.bar(team_appearances, x='team_name', y='count_matches', title='Number of Appearances by Team')
fig.update_layout(xaxis_title='Team', yaxis_title='Number of Appearances')

# Show the plot
fig.show()

###Rank by the number of qualifications Times

In [ ]:
team_unique_tournaments = qualifications_data.groupby('team_name')['tournament_id'].nunique().reset_index()
team_unique_tournaments = team_unique_tournaments.rename(columns={'tournament_id': 'unique_tournaments'})

team_unique_tournaments = team_unique_tournaments.sort_values('unique_tournaments', ascending=False)

fig = px.bar(team_unique_tournaments, x='team_name', y='unique_tournaments', title='Number of Unique Tournament Appearances by Team')
fig.update_layout(xaxis_title='Team', yaxis_title='Number of Unique Tournaments Appearances')

fig.show()

###Number of titles for each Continents

In [ ]:
merged_data = pd.merge(tournaments_data, teams_data, left_on='winner', right_on='team_name', how='left')

winner_count = merged_data['confederation_name'].value_counts().reset_index()
winner_count.columns = ['confederation_name', 'winner_count']

fig = px.bar(winner_count, x='confederation_name', y='winner_count', title='Number of Winners by Confederation')
fig.update_layout(xaxis_title='Confederation', yaxis_title='Number of Winners')

fig.show()

###Association Between the number of goals and winning the Trophy

##Players

###Top players by the number of tournaments apperances

In [ ]:
player_tournament_counts = players_data.groupby(['player_id','full_name'])['count_tournaments'].sum().reset_index()

top_players = player_tournament_counts.sort_values('count_tournaments', ascending=False).head(10)


fig = px.bar(top_players, x='full_name', y='count_tournaments', title='Top Players by Count of Tournaments')
fig.update_layout(xaxis_title='Player Name', yaxis_title='Number of Tournaments')

fig.show()


I wanted to merge the booking dataset with the refrees and see how they associate but I Ran out of time -_-